Bugfix: OBOImporter.inc: Search before retrieving

srobb1 opened the pull request:
# Issue 1) Doing a search for ontology terms before retrieving term


Hello everyone.
I am trying a new install and load with Tripal 3, v 7.x-3.7.    


I loaded go.obo as a test, all is fine.  

 I loaded [plana.obo](https://raw.githubusercontent.com/obophenotype/planaria-ontology/master/plana.obo) and I am getting errors in the log.  
Some of the urls that the loader is using for checking for data are not formed correctly. Is this an issue someone else has seen?   

here is the error I am getting:
```Performing EBI OLS Lookup for: BFO:0000062
Error: no data with http://www.ebi.ac.uk/ols/api/ontologies/bfo/terms/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FBFO_0000062.
```


The url formed has two problems:
   - BFO:0000062 is not [Term] it is [Typedef] 
   - and it is belongs to OVAE not BFO. 
   
 These bits of info are impossible to get from just the ID or iri.

This is the url that would work with the api since it is from ovae and is a property not a term (the bfo is changed to ovae and terms is changed to properties):
https://www.ebi.ac.uk/ols/ontologies/ovae/properties?iri=http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FBFO_0000062
This url works better https://www.ebi.ac.uk/ols/api/search?q=BFO:0000062  

Also something else I am seeing is that some terms in the api call are being compounded:
This particular issue was reported in #708 my code fixes this as well.
```
Performing EBI OLS Lookup for: RO:0001001
Error: no data with http://www.ebi.ac.uk/ols/api/ontologies/ro/terms/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FBFO_00000500001001. 
```
Why is BFO_0000050 being merged with 0001001 of RO:0001001?   

RO:0001001 is also a **property** like BFO:0000062 


I wrote code to performed a search api call and retrieve the top hit and use that iri to retrieve the term information. I no longer see either of the errors reported above.



# issue 2) GFF loader Issues   
See issue #1215


Next I moved on to trying to load my gff file. We have transcripts but no contigs. My transcripts are my landmark features. My Ontology_terms= and Name= where not being inserted into chado. I fixed this by commenting out the if statement on line:1545

Then I noticed that the name was not being updated if it was changed and reimported. I added a updateFeatureName function. I am not sure if this should be more general or if a separate update should be made for each piece that needs to be updated. I would like to incorporate an update for is_obsolet. This is something I am not currently using but plan on using in the future when a version of my transcripts or mRNAs are changed. i want to make a feature obsolete and add a new feature.

# issue 3) Sequence display issues 

Now I see my transcript sequences on my transcript pages. I have two, a derived from landmark in addition to the sequence I manually uploaded with the fasta loader. Next I want to figure out how to disable this in the display on the transcript page. I don't want to see both, especially since they are identical.  

I have disabled the derived from srcfeature functionality on line 136 of [tripal_chado/includes/TripalFields/data__sequence_record/data__sequence_record.inc](https://github.com/tripal/tripal/pull/1228/files#diff-ea6731947cb4ceeb3a3d9822910b1261b494cf94da06990d4cc1597bc81cc348) but checking to see if the feature_id != srcfeature_id

===

For now, I done with bugs and new features, well i fixed all the ones I have come up with. i will continue loading and customizing my T3 site to see what else I come across. I am been busy with other projects for a while now, I am happy to be back on the tripal train!! I want to help with anything I can fit into my to do list!! I would love to work with someone to better test these changes and get them into a format that might be more suitable for the code base. I have done my best to try to understand the code and to mimic coding style.  

Sofia

srobb1 committed:
Update OBOImporter.inc

Doing a search before retrieving term

srobb1 committed:
Update OBOImporter.inc

fixed syntax error on line 1081

srobb1 committed:
Update GFF3Importer.inc

just wnat to find the doc for testing, it is in the pull request form

srobb1 committed:
Update GFF3Importer.inc

undoing my tmp changes to get doc

srobb1 committed:
runs now, searches then gets term

srobb1 committed:
cache lanmarks to add props and update feature names

srobb1 committed:
dont derive sequence from src when you are src

spficklin commented on tripal_chado/includes/TripalImporter/OBOImporter.inc:
```diff
@@ -1040,17 +1040,24 @@ class OBOImporter extends TripalImporter {
+      $query_count++;
+      if (!$ontology_results) {
+        $ontology_results=  $this->oboEbiLookup($id,'query-non-local'); 
+	print('query-non-local');
```
Is this print statement left over debugging that needs to come out?

srobb1 replied:
Left over. Fixed.

spficklin commented on tripal_chado/includes/TripalImporter/OBOImporter.inc:
```diff
@@ -1040,17 +1040,24 @@ class OBOImporter extends TripalImporter {
-      if (!$response) {
-        throw new Exception(t('Did not get a response from EBI OLS trying to lookup ontology: !ontology',
-          ['!ontology' => $short_name]));
+      $query_count = 0;   
```
Can we take out $query_count?  It's not necessary other than in the log message below and it will always be 2 if it makes it to that point, right?

srobb1 replied:
Gone.

spficklin replied:
Did you forget to push a commit?

spficklin commented on tripal_chado/includes/TripalImporter/OBOImporter.inc:
```diff
@@ -1063,25 +1070,61 @@ class OBOImporter extends TripalImporter {
-      $base_iri = $ontology_results['config']['baseUris'][0];
-      $ontologyID = $ontology_results['ontologyId'];
-      $this->baseIRIs[$short_name] = [$ontologyID, $base_iri];
+      # this is good, but i think i need to do the decode
```
Drupal coding style requires comments begin with forward slashes and have properly English grammer (i.e. first letter capitalized, ending the sentence with a period).  Can you fix?

srobb1 replied:
Fixed.

spficklin commented on tripal_chado/includes/TripalImporter/OBOImporter.inc:
```diff
@@ -1063,25 +1070,61 @@ class OBOImporter extends TripalImporter {
+        $obo_id = $each['obo_id'];
+        $defining_ontology= $each['is_defining_ontology'];
+	# first defining_ontology shiould be true, but if it isnt, take the one with the obo_id=$id
+	if ($defining_ontology == 'false' and $obo_id != $id){
```
The indentation is off on this bit of code.  Perhaps your editor is mixing tabs and spaces?  Sorry to be pedantic on this. It's Drupal coding standards.

srobb1 replied:
Fixed. I think I found all my miss-indentations. My vi editor on this particular machine isn't configured as I like and it inserts tabs when I return.

spficklin commented on tripal_chado/includes/TripalImporter/OBOImporter.inc:
```diff
@@ -1063,25 +1070,61 @@ class OBOImporter extends TripalImporter {
+    if ($type == 'class'){
+      $query = 'term';
+    }
+    $results =  $this->oboEbiLookup($id,$query);
```
Another pedantic change.... Drupal coding standards require a space after commas in any list or function arguments.

srobb1 replied:
Got it!

spficklin commented on tripal_chado/includes/TripalImporter/OBOImporter.inc:
```diff
@@ -1063,25 +1070,61 @@ class OBOImporter extends TripalImporter {
+      	    $iri = urlencode(urlencode($each['iri']));
+            $ontology=$each['ontology_name'];
+            $type=$each['type']; # property|class
+	    if ($type == 'property'){
+              $type = 'properties';
+            }elseif($type == 'class'){
+              $type = 'terms';
+            }
+ 	    $full_url = 'http://www.ebi.ac.uk/ols/api/ontologies/' . $ontology . '/' . $type . '/' . $iri;
```
Indention again.

srobb1 replied:
Fixed.

spficklin commented on tripal_chado/includes/TripalImporter/OBOImporter.inc:
```diff
@@ -1063,25 +1070,61 @@ class OBOImporter extends TripalImporter {
-        '!url' => $full_url,
+      $query = 'query-non-local';
+      $ontology_results = $this->oboEbiLookup($id,$query);
+      if($ontology_results ){
```
Drupal coding standards requires a space in control statements before opening parens and none before the closing paren.

```
if ($ontology_results) {
```

srobb1 replied:
Fixed.

spficklin commented on tripal_chado/includes/TripalImporter/OBOImporter.inc:
```diff
@@ -1063,25 +1070,61 @@ class OBOImporter extends TripalImporter {
+      }
+    }
+    if (!$results) {
+                throw new Exception(t('Did not get a response from EBI OLS trying to lookup: !type !id with !url',
+          ['!type'=> $type,'!id' => $id, '!url' => $full_url]));
+      $this->logMessage('Error: no data with !id.  The response was: !response result was:!results', [
+        '!id' => $id,
         '!response' => $response,
+    	'!results' => $results,
```
Indention here too.

srobb1 replied:
Done.

spficklin commented on tripal_chado/includes/TripalImporter/OBOImporter.inc:
```diff
@@ -1063,25 +1070,61 @@ class OBOImporter extends TripalImporter {
+
+            $options = [];
+            $response = drupal_http_request($full_url, $options);
+	    $results = drupal_json_decode($response->data);
```
indention

srobb1 replied:
Done.

spficklin commented on tripal_chado/includes/TripalImporter/OBOImporter.inc:
```diff
@@ -1063,25 +1070,61 @@ class OBOImporter extends TripalImporter {
+            }elseif($type == 'class'){
+              $type = 'terms';
+            }
+ 	    $full_url = 'http://www.ebi.ac.uk/ols/api/ontologies/' . $ontology . '/' . $type . '/' . $iri;
+
+            $options = [];
+            $response = drupal_http_request($full_url, $options);
+	    $results = drupal_json_decode($response->data);
```
I think these lines should go in the `oboEbiLookup` function  and  in that function, add a new `ifelse` statement for "type".  Then these lines get replaced with the function call.

srobb1 replied:
Okay. I created a new ifelse statement and have oboEbiLookup accept a complete iri if one is found in the query. I call the function like this: `$results = $this->oboEbiLookup($id, $type, $found_iri);`

spficklin commented on tripal_chado/includes/TripalImporter/GFF3Importer.inc:
```diff
@@ -615,6 +623,9 @@ class GFF3Importer extends TripalImporter {
       $this->logMessage("Step  8 of 26: Processing !num_features features...               ",
           ['!num_features' => number_format(count(array_keys($this->features)))]);
       $this->insertFeatures();
+      $this->logMessage("Step  8.1 of 26: Processing !num_features feature Names to update...               ",
```
I know this is a pain but can you change this to step 9 and increment those that follow?

srobb1 replied:
Oh Yes!! I meant to do that from the start and got over excited when it all worked!! Fixed.

spficklin commented on tripal_chado/includes/TripalImporter/GFF3Importer.inc:
```diff
@@ -486,6 +493,7 @@ class GFF3Importer extends TripalImporter {
 
     // Set the private member variables of this class using the loader inputs.
     $this->organism_id = $arguments['organism_id'];
+    $this->update_names=[];
```
This line isn't necessary because this variable is set to an empty array on line 112

srobb1 replied:
Fixed.

spficklin commented on tripal_chado/includes/TripalImporter/GFF3Importer.inc:
```diff
@@ -1531,9 +1541,10 @@ class GFF3Importer extends TripalImporter {
+#    if (strcmp($gff_feature['uniquename'], $gff_feature['landmark']) != 0) {
       $this->cacheFeature($gff_feature);
-    }
+#    }
```
Can you remove the inline comments  in the lines above?

srobb1 replied:
Done.

spficklin commented on tripal_chado/includes/TripalImporter/GFF3Importer.inc:
```diff
@@ -1840,6 +1851,58 @@ class GFF3Importer extends TripalImporter {
+
+    $this->setItemsHandled(0);
+    $this->setTotalItems($num_batches);
+#batch update: https://www.alibabacloud.com/blog/how-does-postgresql-implement-batch-update-deletion-and-insertion_596030
```
indentation and change to two forward slashes.

srobb1 replied:
Fixed.

spficklin commented on tripal_chado/includes/TripalImporter/GFF3Importer.inc:
```diff
@@ -1881,6 +1944,13 @@ class GFF3Importer extends TripalImporter {
               if ($matched_type_id == $f->type_id and $matched_organism_id == $f->organism_id) {
                 $this->features[$f->uniquename]['feature_id'] = $f->feature_id;
+                $this->features[$f->uniquename]['name'] = $f->name;
+		# SOFIA: check to see if the name has changed
+		if ($f->name != $matched_feature['name']) {
+                  ## need to update name of feature
+		  ## add flag to cached feature that indicates updated needed
+		  $this->update_names[$f->feature_id]=$matched_feature['name'];
+	        }
```
indentation, and change comments to forwards slashes

srobb1 replied:
Done.

spficklin commented on tripal_chado/includes/TripalImporter/GFF3Importer.inc:
```diff
@@ -2821,6 +2893,7 @@ class GFF3Importer extends TripalImporter {
     $name = '';
 
     // If there is no ID or name then try to create a name and ID.
+    // SOFIA: Name typo?? should name be Name?
     if (!array_key_exists('ID', $attrs) and !array_key_exists('name', $attrs)) {
```
Good catch! Yes, it should be Name (capitalized).

srobb1 replied:
Changed 'name' to 'Name'.

spficklin commented on tripal_chado/includes/TripalFields/data__sequence_record/data__sequence_record.inc:
```diff
@@ -127,10 +127,20 @@ class data__sequence_record extends ChadoField {
       $featureloc = $this->getFeatureLoc($seq['featureloc_id']);
       $coords = $this->getSequenceCoords($featureloc);
 
+      # SOFIA: if the feature is a srcfeature don't derive sequence from itself
+      if ($feature->feature_id !=  $featureloc->srcfeature_id){
+        # i am not quite sure what the $option is doing, i am going to set it to 0 since there is no seq derived from parent
+        $options = [
+          'derive_from_parent' => 0,
+         ];
+         continue;
+      }
+
```
I'm good with this. Can you fix the comment style so #'s are two forwards slashes?  (Drupal coding style)

srobb1 replied:
Done.

spficklin commented:
Thanks @srobb1 for the fixes!   I did a code review and they look good!  Most of my requests for changes are style and formatting but otherwise I think this will fix the issues. If you can make the fixes then I'll do the functional test next

srobb1 committed:
requested corrections complete

srobb1 committed:
made requested corrections

srobb1 committed:
made requested corrections

srobb1 commented:
I hope I didn't break anything when I made these corrections :/ Thank you for pointing out drupal coding format standards!! Good things for me to learn.

srobb1 committed:
missed this one about where the landmark is not cached

spficklin commented on tripal_chado/includes/TripalImporter/OBOImporter.inc:
```diff
@@ -1063,27 +1068,63 @@ class OBOImporter extends TripalImporter {
+        $defining_ontology = $each['is_defining_ontology'];
+        // First result should have defining_ontology=true, but if 
+        // it doesn't, use the first result with obo_id=$id
+	if ($defining_ontology == 'false' and $obo_id != $id) {
+	  continue;
+	}
```
oops another indention issue.

srobb1 replied:
I delete the white space and tried again. I set up my .vimrc. No more crazy white spaces for me.

spficklin commented on tripal_chado/includes/TripalImporter/OBOImporter.inc:
```diff
@@ -1063,27 +1068,63 @@ class OBOImporter extends TripalImporter {
+          if ( !$defining_ontology and $obo_id != $id ) {    
+            continue;
+          }  
+      	  $found_iri = urlencode(urlencode($each['iri']));
+          $ontology = $each['ontology_name'];
+	  // Type should be 'property' or 'class' in the response
+          $type = $each['type'];
+	  if ($type == 'property') {
+            $type = 'properties';
+	  }
+	  elseif ( $type == 'class') {
+            $type = 'terms';
+          }
+	  $results = $this->oboEbiLookup($id, $type, $found_iri);
+        } 
+      }
```
More indention issues.

srobb1 replied:
Weird, it "looks" fine in my editor. i will delete the all the white space and try again.

spficklin commented on tripal_chado/includes/TripalImporter/OBOImporter.inc:
```diff
@@ -2553,6 +2604,14 @@ class OBOImporter extends TripalImporter {
       }
     }
+    elseif ($found_iri) {
+      $full_url = 'http://www.ebi.ac.uk/ols/api/ontologies/' . $ontology . '/' . $type . '/' . $found_iri;
```
I think this may be a problem as `$type` isn't defined in this function....  

Also, maybe I confused things a bit.  The $found_iri variable?  An IRI would a full path to an object (e.g. a full URL) so the variable name confuses me here. What is in the $found_uri variable?

srobb1 replied:
in line 1049 I run a query to find the term `$ontology_results =  $this->oboEbiLookup($id, 'query');`
then there are checks. In line 1072 I go through the query results with a foreach `foreach ($ontology_results['response']['docs'] as $each ) {`. I look at each result to see if it is what we want, it should be defining_ontology=true, if not, it should have the same ID. 

```
      foreach ($ontology_results['response']['docs'] as $each ) {
        $obo_id = $each['obo_id'];
        $defining_ontology = $each['is_defining_ontology'];
        // First result should have defining_ontology=true, but if
        // it doesn't, use the first result with obo_id=$id
        if ($defining_ontology == 'false' and $obo_id != $id) {
          continue;
        }
        $iri = urlencode(urlencode($each['iri']));
        $ontologyID = $each['ontology_name'];
        // Type should be 'property' or 'class' in the response
        $type = $each['type'];
      }
    }
```
Then if it is a good hit in our query results, collect the iri, I call it $found_iri, so that it is more clear that it was retrieved from the query and not built using the accession and an ontology prefix. Since sometimes (though not too often) the iri cannot be formed some cleanly.

Does that make more sense? I query the accession, get hits, pick a good hit, get the good hit iri, then use the found iri in the term api call.

spficklin replied:
Yes, but does the $found_iri contain a full URL?   

Also, I think the $found_iri variable should have a NULL default in the function arguments so that other function calls that are made that don't break.

srobb1 replied:
Also. I altered the oboEbiLookup to take the found_iri since it could differ from a built iri. If the found_iri is defined then it means we should use that iri in the api call.

srobb1 replied:
Ok. $type was a leftover from when this code was in the main. I addressed this now with checking for type and formatting the correct text for the api url.

srobb1 replied:
$found_iri will contain the values found in the 'iri' key: `"iri": "http://purl.obolibrary.org/obo/RO_0002206",` which is  a full iri.

See this sample api search:
https://www.ebi.ac.uk/ols/api/search?q=RO:0002206&queryFields=obo_id

srobb1 replied:
I am setting the default of $found_iri = NULL in the  oboEbiLookup function. is this the right syntax? `private function oboEbiLookup($accession, $type_of_search, $found_iri = NULL) {` Is it null or NULL?

spficklin replied:
Yes that's right.  I'll go ahead and test.

spficklin commented:
Thanks @srobb1 .  I added a few more formatting issues and also one question about the new code in the `oboEbiLookup` function

srobb1 committed:
think i got the weird white space and the query_count

srobb1 commented:
I need to look at the code in the webbrowser to see if the white space looks better

srobb1 committed:
white space fixed

srobb1 committed:
more white space fixed

srobb1 committed:
more white space fixed

srobb1 committed:
more white space

srobb1 committed:
addressed $type

srobb1 committed:
added default null

spficklin commented:
@srobb1 to help me out with testing, can you share a GFF3 file (via slack if you like)?

spficklin commented:
@srobb1 when  I import PLANO I originally thought it worked but it didn't seem to import any terms and failed with this message:

```
Did not get a response from EBI OLS trying to lookup: property BFO:0000062 with 
[site http://default] [TRIPAL ERROR] [TRIPAL_JOB] Did not get a response from EBI OLS trying to lookup: property BFO:0000062 with
```
Which is the term 'preceded by' and is a property. So I think something might still be awry.  When you try to load the file do you get the same problem?

srobb1 commented:
@spficklin Hmmmm.  I have it in my DB. I had the problem at first, but when I changed the code it loaded okay, but I did not start from scratch again. Ok. I need to start from scratch, load with this code, and do some more testing print statements. I am wondering what the url that was used to try to get that term info.  BFO:0000062 is definitely a property. I wonder if I messed something up when I moved those few lines into the lookup function and the correct else if statement is not being used.

```
drupal=# select *, name from cvterm where name like '%preced%';
 cvterm_id | cv_id |            name             | definition | dbxref_id | is_obsolete | is_relationshiptype |            name
-----------+-------+-----------------------------+------------+-----------+-------------+---------------------+-----------------------------
     50358 |    99 | preceded by                 |            |     78813 |           0 |                   1 | preceded by

drupal=# select * from dbxref where dbxref_id = 78813;
 dbxref_id | db_id | accession | version | description
-----------+-------+-----------+---------+-------------
     78813 |    46 | 0000062   |         |
(1 row)

drupal=# select * from db where db_id = 46;
 db_id | name | description | urlprefix | url
-------+------+-------------+-----------+-----
    46 | BFO  |             |           |
(1 row)

```

srobb1 committed:
needed to use found ontology. forgot to pass this to function when rearranging

srobb1 commented:
When I moved those lines of code from the main to the oboEbiLookup function, I forgot that I also need to pass in the found ontology. The issue with that BFO terms is that it is not identified as coming from BFO, but from OVAE. We got errors because it was using bfo in the api call url. I hope I don't have any more spacing issues. I will check again. I got a new aws instance started up and my vimrc was not mine again. I should have just added my vimrc from the start.

srobb1 commented:
Also I had to move my found_iri elseif block up so that the elseif type is properties did not catch my cases.

srobb1 committed:
correct Step 21 naming

spficklin commented:
Thank you very much @srobb1 for taking time to dig into these loaders, which are not the easiest bit of code to go  through.  The fixes worked great after all the additional fixes.
